core/src/main/java/edu/wpi/grip/core/metrics/BenchmarkRunner.java

```diff
@@ -0,0 +1,104 @@
+package edu.wpi.grip.core.metrics;
+
+import edu.wpi.grip.core.events.BenchmarkEvent;
+import edu.wpi.grip.core.events.BenchmarkRunEvent;
+import edu.wpi.grip.core.events.RunStartedEvent;
+import edu.wpi.grip.core.events.RunStoppedEvent;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Benchmark runner. This runs the pipeline multiple times to collect data about how long each
+ * step takes to run.
+ */
+@Singleton
+public class BenchmarkRunner {
+
+  private EventBus eventBus;
+
+  private AtomicBoolean isBenchmarking = new AtomicBoolean(false);
+  private AtomicInteger runsRemaining = new AtomicInteger(0);
+
+  @Inject
+  BenchmarkRunner(EventBus eventBus) {
+    this.eventBus = eventBus;
+  }
+
+  /**
+   * Benchmarks the pipeline.
+   *
+   * @param numRuns the number of runs of the pipeline to collect events from.
+   */
+  public void run(int numRuns) {
+    checkArgument(numRuns > 0, "Must run the pipeline at least once");
+    this.isBenchmarking.set(true);
+    this.runsRemaining.set(numRuns);
+    eventBus.post(BenchmarkEvent.started());
+    eventBus.post(new BenchmarkRunEvent());
```

Comment:
Basic event flow:

1. BenchmarkEvent.started()
2. BenchmarkRunEvent
3. RunStartedEvent
4. RunStoppedEvent
5. Repeat 2-4 until the number of runs have completed (or it was cancelled)
6. BenchmarkEvent.stopped()